### PR TITLE
[SEDONA-663] Support spark connect python api

### PR DESCRIPTION
**Pull request by sebbegg**
## Did you read the Contributor Guide?

- Yes, I have read the [Contributor Rules](https://sedona.apache.org/latest-snapshot/community/rule/) and [Contributor Development Guide](https://sedona.apache.org/latest-snapshot/community/develop/)

## Is this PR related to a JIRA ticket?

- Yes: https://issues.apache.org/jira/browse/SEDONA-663

## What changes were proposed in this PR?
This PR enables the python dataframe api to be used in spark-connect sessions. See https://github.com/apache/sedona/issues/1619 for the current status.

## How was this patch tested?
* Ran `tests/sql/test_dataframe_api.py` locally against a (dockerized) spark connect session using sedona 1.6.1.
* Successfully ran the python build workflow for spark 3.5

## Did this PR include necessary documentation updates?

- No, this PR does not affect any public API so no need to change the documentation.

<sub>Sebastian Eckweiler (sebastian.eckweiler@mercedes-benz.com), Mercedes-Benz AG on behalf of MBition GmbH ([More Info](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md))</sub>

### CODE REVIEW

File: python/sedona/sql/dataframe_api.py [hunk 1]
```diff
@@ -49,13 +64,15 @@ def call_sedona_function(
...
-    if (
-        (not isinstance(args, Iterable))
-        or isinstance(args, str)
-        or isinstance(args, Column)
+    if (not isinstance(args, Iterable)) or isinstance(
+        args, (str, Column, ConnectColumn)
     ):
         args = [args]
 
+    # in spark-connect environments use connect api
```
Comment (jbampton) [inline python/sedona/sql/dataframe_api.py:72]:
```suggestion
    # in spark-connect environments use connect API
```

File: python/sedona/spark/SedonaContext.py [hunk 2]
```diff
@@ -34,8 +41,11 @@ def create(cls, spark: SparkSession) -> SparkSession:
         :return: SedonaContext which is an instance of SparkSession
         """
         spark.sql("SELECT 1 as geom").count()
-        PackageImporter.import_jvm_lib(spark._jvm)
-        spark._jvm.SedonaContext.create(spark._jsparkSession, "python")
+
+        # with spark connect there is no local jvm
```
Comment (jbampton) [inline python/sedona/spark/SedonaContext.py:45]:
```suggestion
        # with Spark Connect there is no local JVM
```

File: .github/workflows/python.yml [hunk 3]
```diff
@@ -153,3 +153,20 @@ jobs:
           SPARK_VERSION: ${{ matrix.spark }}
           HADOOP_VERSION: ${{ matrix.hadoop }}
         run: (export SPARK_HOME=$PWD/spark-${SPARK_VERSION}-bin-hadoop${HADOOP_VERSION};export PYTHONPATH=$SPARK_HOME/python;cd python;pipenv run pytest tests)
+      - env:
+          SPARK_VERSION: ${{ matrix.spark }}
+          HADOOP_VERSION: ${{ matrix.hadoop }}
+        run: |
+          if [ ! -f "spark-${SPARK_VERSION}-bin-hadoop${HADOOP_VERSION}/sbin/start-connect-server.sh" ]
+          then
+            echo "Skipping connect tests for spark $SPARK_VERSION"
```
Comment (jbampton) [inline .github/workflows/python.yml:162]:
```suggestion
            echo "Skipping connect tests for Spark $SPARK_VERSION"
```

File: python/sedona/sql/dataframe_api.py [hunk 4]
```diff
@@ -24,8 +24,23 @@
 from pyspark.sql import Column, SparkSession
 from pyspark.sql import functions as f
 
-ColumnOrName = Union[Column, str]
-ColumnOrNameOrNumber = Union[Column, str, float, int]
+try:
+    from pyspark.sql.connect.column import Column as ConnectColumn
+    from pyspark.sql.utils import is_remote
+except ImportError:
+    # be backwards compatible with spark < 3.4
```
Comment (jbampton) [inline python/sedona/sql/dataframe_api.py:31]:
```suggestion
    # be backwards compatible with Spark < 3.4
```

File: python/sedona/sql/dataframe_api.py [hunk 5]
```diff
@@ -86,6 +103,10 @@ def _get_type_list(annotated_type: Type) -> Tuple[Type, ...]:
     else:
         valid_types = (annotated_type,)
 
+    # functions accepting a Column should also accept the spark connect sort of Column
```
Comment (jbampton) [inline python/sedona/sql/dataframe_api.py:106]:
```suggestion
    # functions accepting a Column should also accept the Spark Connect sort of Column
```

**Comment by sebbegg** [comment]
Switchd to use `UnresolvedFunction` instead of `CallFunction` which is only available for spark >= 3.5.
`UnresolvedFunction` seems "stable" in that it exists in spark >= 3.4, including the 4.0 preview releases.

**Review by jiayuasu — APPROVED** [review]
Great work. Thank you!

**Pull request merged**